### ISSUE TITLE
Reproducing zero-shot classification results

### ISSUE THREAD

**Issue by Hasindri** [wisdomikezogwo/quilt1m#10]
Hi, thank you very much for this great work on image-text contrastive training for histopathology and also publishing a valuable dataset.

I used the provided pre-trained QuiltNet along with given tokenizer to reproduce zero-shot classification results on NCT-CRC-HE-100k dataset. Used following commands,

```
import open_clip

model, preprocess_train, preprocess_val = open_clip.create_model_and_transforms('hf-hub:wisdomik/QuiltNet-B-32')
tokenizer = open_clip.get_tokenizer('hf-hub:wisdomik/QuiltNet-B-32')
```

I also used the class names and templates as given in the paper as follows,

```


nct_classnames = ["Adipose", "Debris", "Lymphocytes", "Mucus", "Smooth muscle", "Normal colon mucosa", "Cancer-associated stroma", "Colorectal adenocarcinoma epithelium"]


nct_template = [
    lambda c: f'a histopathology slide showing {c}.',
    lambda c: f'histopathology image of {c}.',
    lambda c: f'pathology tissue showing {c}.',
    lambda c: f'presence of {c} tissue on image.',
]
 ```

But I get a top1 accuracy lower than what's reported in the paper (59.56%), I get 

```
zero shot metrics {'nct-zeroshot-val-top1': 0.28518236912136324, 'nct-zeroshot-val-top5': 0.7248697363418835}

``` 

I also tried training my own QuiltNet using Open_clip codebase from OpenAI, and the results were,

```
zero shot metrics {'nct-zeroshot-val-top1': 0.30728805599660086, 'nct-zeroshot-val-top5': 0.6808149026097458}
```
Could you kindly help me understand why I am. unable to reproduce the given numbers? I need to understand what I might be doing wrong.

Thank you.

**Comment by wisdomikezogwo** [comment]
Hi,

I'm not quite sure where the issue might be, I just re-ran our evaluations to be sure. However, I have put together a quick code-snippet you could use to evaluate both models, and report back your findings.

Assuming you have your dataset class in a dictionary (`data_dict`) like this:
```
{'nck': Dataset ImageFolder
     Number of datapoints: 6333
     Root location: /path to nck test set
     StandardTransform
 Transform: Compose(
                Resize(size=224, interpolation=bicubic, max_size=None, antialias=None)
                CenterCrop(size=(224, 224))
                <function _convert_to_rgb at 0x7fc544e8e550>
                ToTensor()
                Normalize(mean=(0.48145466, 0.4578275, 0.40821073), std=(0.26862954, 0.26130258, 0.27577711))
            )}
```
            
Try using this function to evaluate the model.

```python
def run_zero_shot(model_name, data_dict, pre_trained_path='', context_lenght=77, wrks=0, bsz=32, device='cpu'):
    model, pct, pv = open_clip.create_model_and_transforms(model_name, pretrained=pre_trained_path)
    tokenizer = open_clip.get_tokenizer(model_name)

    model.to(device)
    model.eval()

    res = {}
    for data_name, data in data_dict.items():  
        labels = data.classes
        print(labels)
        
        templates = ["a histopathology slide showing {c}",
            "histopathology image of {c}",
            "pathology tissue showing {c}",
            "presence of {c} tissue on image"]
        
        if pv:
            data.transform = pv

        loader = torch.utils.data.DataLoader(data, batch_size=bsz, 
                                             shuffle=True, num_workers=wrks,
                                             pin_memory=False
                                            )

        pred = []
        true = []

        for images, target in tqdm(loader):
            images, target = images.to(device), target.to(device)

            with torch.no_grad(): 
                image_features = model.encode_image(images, normalize=True)

                zeroshot_weights = []
                for classname in labels:
                    texts = [template.format(c=classname) for template in templates]
                    texts = tokenizer(texts).to(device)  # tokenize
                    class_embeddings = model.encode_text(texts)
                    class_embedding = F.normalize(class_embeddings, dim=-1).mean(dim=0)
                    class_embedding /= class_embedding.norm()
                    zeroshot_weights.append(class_embedding)
                    
                zeroshot_weights = torch.stack(zeroshot_weights, dim=1).to(device)
                logits = (100. * image_features @ zeroshot_weights).detach()#.softmax(dim=-1)

                true.append(target.cpu())
                pred.append(logits.float().cpu())

        pred = torch.cat(pred)
        true = torch.cat(true)

        topk = (1,)
        pred_n = pred.topk(max(topk), 1, True, True)[1].t()
        correct = pred_n.eq(true.view(1, -1).expand_as(pred_n))
        n = len(true)
        res[data_name] = [float(correct[:k].reshape(-1).float().sum(0, keepdim=True).cpu().numpy()) / n for k in topk][0]
        print(res)
    return res
```
by running this to get the top 1 ACC value:

```python
model_name = "hf-hub:wisdomik/QuiltNet-B-32"
results = run_zero_shot(model_name, data_dict, pre_trained_path='', wrks=0, bsz=256, device=device)
```

This is what I get for NCK
`['adipose', 'debris', 'lymphocytes', 'mucus', 'smooth muscle', 'normal colon mucosa', 'cancer-associated stroma', 'colorectal adenocarcinoma epithelium']
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:41<00:00,  1.65s/it]
{'nck': 0.5956102952786989}
`

**Comment by Hasindri** [comment]
Thank you very much for the quick response. Now I am able to reproduce the same result as you when using the QuiltNet-B-32 model from hugging face models that you provided. However, I trained a CLIP initialised ViT-B-32 & GPT(77) model on quilt-1M by modifying official open_clip codebase https://github.com/mlfoundations/open_clip. When using that pretrained checkpoint I get 

```
{'nct': 0.29685772935417654}
```

Could you kindly share the training script for QuiltNet or the hyper-parameters  you used (I utilised default ones from open_clip model) and also for how many epochs did you do contrastive language-image pretraining on quilt-1M?

**Comment by wisdomikezogwo** [comment]
Hi,

We also use the ope_clip training codebase and only make the changes highlighted in the C2 section of the appendix in the paper. We employ you to try finetuning for a few epochs first, say 4 epochs, and iteratively test performance.

**Comment by Hasindri** [comment]
Hi @wisdomikezogwo, I experimented as you mentioned. Used following parameters (as same as what's mentioned in appendix of your paper, except for a 512 batch size instead of 1024) and ran the main.py training script in open_clip codebase.

```
C: 3.16
aggregate: True
batch_size: 512
beta1: 0.9
beta2: 0.98
checkpoint_path: ./logs/lr=0.0005_wd=0.2_agg=True_model=ViT-B/32_batchsize=512_workers=1_date=2023-09-07-09-03-20/checkpoints
copy_codebase: False
csv_caption_key: caption
csv_img_key: image_path
csv_separator: 	
dataset_type: auto
debug: True
dist_backend: nccl
dist_url: tcp://127.0.0.1:6100
distributed: True
dp: False
epochs: 40
eps: 1e-06
gpu: 0
imagenet_v2: None
imagenet_val: None
log_level: 10
log_path: ./logs/lr=0.0005_wd=0.2_agg=True_model=ViT-B/32_batchsize=512_workers=1_date=2023-09-07-09-03-20/out.log
logs: ./logs/
lr: 0.0005
model: ViT-B/32
multigpu: None
name: lr=0.0005_wd=0.2_agg=True_model=ViT-B/32_batchsize=512_workers=1_date=2023-09-07-09-03-20
nct: None
ngpus_per_node: 4
openai_pretrained: True
precision: amp
rank: 0
regression_frequency: 2
report_to: 
resume: None
save_frequency: 1
skip_aggregate: False
skip_scheduler: False
tensorboard: False
tensorboard_path: 
train_data: /data1/dri/open_clip/data/quilt1m/quilt_1M_lookup_updated.csv
use_bn_sync: False
val_data: None
wandb: False
wandb_notes: 
warmup: 2000
wd: 0.2
workers: 1
world_size: 4
zeroshot_frequency: 2
```

I used openAI pretrained weights to initialise and trained the model on quilt-1M for 40 epochs, saved the checkpoint at each epoch and did zero shot evaluation on NCT dataset (CRC-HE-7k to be exact), following are the results,

```


Epoch | Top-1 Accuracy

1 | 0.476235591346913
2 | 0.49265750828991
3 | 0.6039791567977260
4 | 0.6682456971419550
5 | 0.587715142902258
6 | 0.5791883783357020
7 | 0.40438970472130100
8 | 0.515079741039002
9 | 0.5092373282804360
10 | 0.5463445444497080
11 | 0.3377546186641400
12 | 0.39081004263382300
13 | 0.5067108795199750
14 | 0.4782883309647880
15 | 0.5191852202747510
16 | 0.5065529764724460
17 | 0.46407705668719400
18 | 0.40928469919469400
19 | 0.4419706300331600
20 | 0.49565766619295800
21 | 0.4225485551871150
22 | 0.4069161534817620
23 | 0.4647086688773090
24 | 0.39033633349123600
25 | 0.28690983735986100
26 | 0.32322753829148900
27 | 0.3822832780672670
28 | 0.34533396494552300
29 | 0.32164850781620100
30 | 0.3789673140691620
31 | 0.37801989578398900
32 | 0.3041212695405020
33 | 0.35528185693983900
34 | 0.32733301752723800
35 | 0.33901784304437100
36 | 0.3494394441812730
37 | 0.33349123638086200
38 | 0.33270172114321800
39 | 0.32812253276488200
40 | 0.3304910784778150

```

I am not sure how to explain the reducing behaviour of classification accuracy.

**Comment by Hasindri** [comment]
@wisdomikezogwo I wanted to confirm whether I am using the proper dataset as same as you used. After getting access to 512 x 512 patched dataset via https://zenodo.org/record/8239942, I cleaned the quilt_1M_lookup.csv to only include entries with valid image paths. It reduced the number of entries from 1000k to ~ 814k (814384 to be exact), so I am actually training on 814384 image-text pairs. Do I have to seperately download PubMed, OpenPath and Laion datasets that you have included in quilt-1M, or are those images already included in your version of the dataset that is publicly available (I downloaded and it has 522569 images) . I am trying to understand the difference between the number of pairs, 814384 I get and the numbers you mentioned (802,148 pairs in Quilt + 59371 PubMed pairs + 22682 LAION pairs + 133511 OpenPath pairs totalling to 1,017,712 pairs). I really appreciate any help on this matter.

Also can you kindly let me know if you trained on a separate train split and not the entire quilt-1M?

**Comment by Hasindri** [comment]
posting this in case it helps anyone who is trying to reproduce, 

I got zero shot eval results on NCT-CRC according to how @wisdomikezogwo suggested, when image encoder is frozen bit in the middle of fine-tuning CLIP on quilt-1M, the ACC progresses better,

```

Epoch | Top-1 ACC(visual encoder frozen in 10-16 epochs) | Top-1 ACC(both encoders fine-tuning all epochs)

1 | 0.49044686562450700 | 0.5321332701721140
2 | 0.558345176061898 | 0.5065529764724460
3 | 0.4906047686720350 | 0.5098689404705510
4 | 0.5037107216169270 | 0.42286436128217300
5 | 0.5048160429496290 | 0.4945523448602560
6 | 0.5147639349439450 | 0.4962892783830730
7 | 0.30980577925154000 | 0.41007421443233900
8 | 0.32843833885994 | 0.5542396968261490
9 | 0.45128690983736000 | 0.5679772619611560
10 | 0.4901310595294490 | 0.4048634138638880
11 | 0.4522343281225330 | 0.5469761566398230
12 | 0.5070266856150320 | 0.4645507658297810
13 | 0.5431864834991320 | 0.535922943312806
14 | 0.36886151902731700 | 0.4370756355597660
15 | 0.41196905100268400 | 0.4942365387651980
16 | 0.5239223117006160 | 0.4942365387651980
17 | 0.4498657824096010 | 0.4288646770882680
18 | 0.5059213642823310 | 0.36933522816990400
19 | 0.46281383230696400 | 0.4361282172745930
20 | 0.505605558187273 | 0.3805463445444500
21 | 0.5229748934154430 | 0.4107058266224540
22 | 0.5392389073109110 | 0.4187588820464240
23 | 0.535922943312806 | 0.4230222643297020
24 | 0.5360808463603350 | 0.3895468182535920
25 | 0.5345018158850470 | 0.3797568293068060

```
